### ISSUE TITLE
Should missing semicolons result in a parse error?

### ISSUE THREAD

**Issue by jeddy3** [shellscape/postcss-less#54]
Firstly, thanks for your work on this parser!

The following code:

``` less
.class5 {
  color: green
  .mixin()
  .mixin2
}
```

Is parsed to:

``` shell
Declaration {
  raws: { before: '\n  ', between: ': ' },
  type: 'decl',
  parent: 
   Rule {
     raws: { before: '', between: ' ', semicolon: false, after: '\n' },
     type: 'rule',
     nodes: [ [Circular] ],
     parent: 
      Root {
        raws: [Object],
        type: 'root',
        nodes: [Object],
        source: [Object],
        lastEach: 36,
        indexes: [Object] },
     source: { start: [Object], input: [Object], end: [Object] },
     selector: '.class5',
     lastEach: 37,
     indexes: { '37': 0 } },
  source: 
   { start: { line: 2, column: 3 },
     input: 
      Input {
        css: '.class5 {\n  color: green\n  .mixin()\n  .mixin2\n}\n',
        file: '/home/sheo13666q/IdeaProjects/stylelint/test/test.less' },
     end: { line: 4, column: 9 } },
  prop: 'color',
  value: 'green\n  .mixin()\n  .mixin2' }
```

i.e. a rule with a single declaration. 

This is also how the standard PostCSS [parses it](https://astexplorer.net/#/wglK87gmzj). However, the Less parser [throws a parse error](http://leafo.net/lessphp/editor.html#.class5%20%7B%0A%20%20color%3A%20green%0A%20%20.mixin%28%29%0A%20%20.mixin2%0A%7D). In stylelint, we're just [going to document](https://github.com/stylelint/stylelint/issues/1608) this limitation but I wanted to create an issue here in case you want, if it's at all possible, to match the Less parser's behaviour in this situation.

**Comment by shellscape** [comment]
@jeddy3 I'm on the job now :)

I'd like to keep in line with PostCSS, but at the same time accommodate folks who are looking for a stricter experience. We could piggy back on PostCSS `options` and provide for a `{ semicolons: 'strict' }` option or some such, which could instruct the syntax to throw on missing semis. Thoughts?

**Comment by jeddy3** [comment]
> I'm on the job now :)

Excellent news! :)

> which could instruct the syntax to throw on missing semis. 

I suspect stylelint would want to throw on missing semis for consistency with the actual Less parser. I'm not sure what the normal convention is here i.e. strict by default or make stricter with options? I often see `loose` options, so perhaps that's the prevailing wisdom i.e. strict by default but provide a `loose` option for allowing missing semis?

**Comment by shellscape** [comment]
In doing work on #70, I found that the code looking for differentiations in declarations/mixins/extends is pretty fragile; it's really only looking for ending-characters like `}` and ';' to determine when a 'node' ends and the next one, or child, begins. I can see now why webschik left it not throwing errors - it'd be pretty hard to parse consistently to determine where missing semicolons should go.

**Comment by jeddy3** [comment]
@shellscape Thanks for digging into it some more!

> it'd be pretty hard to parse consistently to determine where missing semicolons should go.

Understood. Thanks for explaining the reasoning for why it behaves this way. We've documented the limitation of the `declaration-block-trailing-semicolon` with Less syntax. So, I think we can close this issue. Thanks again for looking into it.

**Comment by shellscape** [comment]
No worries, appreciate the discussion!
